### DISCUSSION TITLE
CLI Client For The Same Functionality

### DISCUSSION THREAD

**Discussion by raveesh-me** [hiranthaR/Json-to-Dart-Model#20] [Ideas]
## Ideal Use Case
As an engineering manager,
I want to maintain a single JSON file in my codebase called `all_models.json` and have build_runner watch it for changes, generate all the model files using `json to dart model` and then generate the `.g.dart` files for json_serializable.

This will improv the cleanliness of my project like crazy..

## Most Immediate Use Case
While the above use case will be crazy good, for now I could get a script to do the `all_models.json` to `json to model` conversion and maybe write a watcher myself to run that script on changes to json file, and the build_runner watcher will then do its thing.
The only missing piece of the puzzle, is a cli interface and a binary for this amazing vscode plugin, but then again if there was a way to use this plugin as a cli interface via a vscode api that I dont know about, that would be a third layer of cool without even bothering you to lift a finger!

**Comment by hiranthaR** [comment]
I will try to do this when I am free. @iamarnas what do you think

**Reply by iamarnas** [reply]
Hi, @raveesh-me @hiranthaR . 
It would make this extension different from the others. And the first idea look like possible fix with small changes. Just create CLI command which generate json objects from the file (like already this extension can grab json from the clipboard) and save to `/models/baseClassName/(generated files...)` code generate function will run `json_serializable` bulder as already suport this function.

**Comment by iamarnas** [comment]
@raveesh-me  HI,
 I added possibility to build all your json models from the one single file. Maybe it match your preferences :) [Read how it works.](https://github.com/hiranthaR/Json-to-Dart-Model/blob/master/README.md#convert-from-file)

**Comment by emileswain** [comment]
Is there a way of runing this directly from a flutter project as a script ? I personaly use Intelij, other developers use VS. So i'm in this position, where I would rather have the script generation, decoupled from the plugin itself. I.e. if it was a npm plugin, i would call be able to call something like `npx json-to-dart-model jsonToDart.fromFile` or something like that, and it would generate the files. 

I was thinking of forking and seeing if it would be a simple call, but it looks like terminal and window, are VS specific elements and coupled into the file generation side of things.

**Reply by iamarnas** [reply]
@emileswain Hi. This extension does not support running directly from a flutter project as a script. The extension depends on VS-specific elements as you see. But it is possible to implement with the help of some npm packages. But I think only a small group of users would benefit from this and I have not thought about :)

**Reply by emileswain** [reply]
hey, coolio, wanted to check first. I looked at other similar solutions and they're not as good or feature complete as yours. So I wanted to check first there wasn't a simple way of doing it I wasn't aware of. I've just about managed to get it to work sans Vscode, so i'll fork the repo if it works.

**Reply by iamarnas** [reply]
@emileswain It's fun if you can make it work without vscode. I was asked many times about Intelij plugin and it should be good news for Intelij users 👍 But if you want to run a simple conversion with this extension is exist already a plugin on Intelij [JsonToDart ​(JSON To Dart)](https://plugins.jetbrains.com/plugin/12562-jsontodart-json-to-dart-). With vscode, this plugin JSON to Dart Model is different from the other existing plugins because it is very customizable, you can add methods like `copyWith`, `toString`, the equality operator. Make enhanced file names like `user.model.dart` to make friendly with other converters such as like Hive Object Converter, rename suffix to the methods `to/from`, and many more... If you can't implement all of the functions by running script. maybe it's not worth twise your time?

**Reply by iamarnas** [reply]
@emileswain It's possible to implement it if conversion runs from the file or directory. Then it is possible to override all options in this way #65 by adding an option separately to every JSON objects :) Then should work with one command and the generator would read options from the file and from every JSON object separately.

**Reply by emileswain** [reply]
oh, great, i'll chek the intellij version as well. 
yeah, i've manage to get it working stripping out the vscode & window imports. which for the most part were used for logging error messages and requesting inputs from the user. 

What i've updated is pretty horible though, would certainly need some work to tweek. The main changes i made were to first remove the calls to window for error messaging. This could be resolved with perhaps an interface so as to be able to switch to console logs for runninng locally.

I then looked for the vscode imports, which tended to be used primarily for requesting user inputs. So i commented out all of those and manually created an InputSettings class with the options. 

So it works, sort off. But if the intellij plugin creates the same output as the VS plugin then i might just stick with that. As long as developers on intellij and vs can replicate each others code. 

I'll fork and push my edits, but it is pretty awfull at the moment.
